src/zcash_script.rs

```diff
@@ -0,0 +1,59 @@
+use super::interpreter::*;
+
+/// This maps to `zcash_script_error_t`, but most of those cases aren’t used any more. This only
+/// replicates the still-used cases, and then an `Unknown` bucket for anything else that might
+/// happen.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[repr(u32)]
+pub enum Error {
+    /// Any failure that results in the script being invalid.
+    Ok = 0,
+    /// An exception was caught.
+    VerifyScript = 7,
+    /// Some other failure value recovered from C++.
+    Unknown(u32),
+}
+
+/// A function which is called to obtain the sighash.
+///    - script_code: the scriptCode being validated. Note that this not always
+///      matches script_sig, i.e. for P2SH.
+///    - hash_type: the hash type being used.
+///
+/// The underlying C++ callback doesn’t give much opportunity for rich failure reporting, but
+/// returning `None` indicates _some_ failure to produce the desired hash.
+///
+/// TODO: Can we get the “32” from somewhere rather than hardcoding it?
+pub type SighashCallback = dyn Fn(&[u8], HashType) -> Option<[u8; 32]>;
```

Comment:
This will also work instead of the generic closure signature.

```suggestion
pub type SighashCallback<'a> = &'a dyn Fn(&[u8], HashType) -> Option<[u8; 32]>;
```

Reply:
This is perfect, thanks!